clarity/contracts/pool/alex-launchpad.clar

```diff
@@ -300,125 +295,135 @@
-          (not (try! (is-listing-completed token)))
-        )
-        (begin
-          (as-contract (unwrap! (contract-call? token-trait transfer-fixed (* (get amount-per-ticket details) ONE_8) tx-sender claimer none) ERR-TRANSFER-FAILED))
-          (as-contract (unwrap! (contract-call? .token-wstx transfer-fixed wstx-per-ticket-in-fixed tx-sender (get fee-to-address details) none) ERR-TRANSFER-FAILED))
-          (as-contract (try! (contract-call? ticket-trait burn-fixed ONE_8 tx-sender)))
-          (map-set listing 
-            token 
-            (merge details 
-              { 
-                last-random: this-random, 
-                tickets-won: (+ tickets-won u1)
-              }
-            )
-          )          
-          (map-set subscriber-at-token 
-            { token: token, user-id: user-id } 
-            (merge sub-details 
-              { 
-                ticket-balance: (- ticket-balance u1), 
-                tickets-won: (+ (get tickets-won sub-details) u1), 
-                wstx-locked-in-fixed: (- (get wstx-locked-in-fixed sub-details) wstx-per-ticket-in-fixed)
-              }
-            )
-          )      
-          (ok true)
-        )
-        (begin
-          (as-contract (unwrap! (contract-call? .token-wstx transfer-fixed (get wstx-per-ticket-in-fixed details) tx-sender claimer none) ERR-TRANSFER-FAILED))
-          (as-contract (try! (contract-call? ticket-trait burn-fixed ONE_8 tx-sender)))
-          (map-set listing 
-            token 
-            (merge details 
-              { 
-                last-random: this-random
-              }
-            )
-          )
-          (map-set subscriber-at-token 
-            { token: token, user-id: user-id } 
-            (merge sub-details 
-              { 
-                ticket-balance: (- ticket-balance u1),
-                tickets-lost: (+ (get tickets-lost sub-details) u1),
-                wstx-locked-in-fixed: (- (get wstx-locked-in-fixed sub-details) wstx-per-ticket-in-fixed)
-              }
-            )
-          )      
-          (ok false)
+          (>= (mod this-random total-subscribed) (get value-low values-adjusted)) 
+          (<= (mod this-random total-subscribed) (get value-high values-adjusted))
+          (not (try! (is-listing-completed (contract-of token-trait))))
         )
+        (you-won token-trait ticket-trait this-random)
+        (you-lost token-trait ticket-trait this-random)
       )
-    )
   )
 )
 
-;; implementation of Linear congruential generator following POSIX rand48
-(define-private (get-next-random (last-random uint))
-    (mod (+ (* u25214903917 last-random) u11) (pow u2 u48))
+;; computes a random number by converting into uint the lower 16 bytes of hash of the sum of ticket position and last random
+(define-private (get-next-random (last-random uint) (ticket-position uint))    
+    (buff-to-uint-le (lower-16-le (sha256 (+ ticket-position last-random))))
```

Comment:
👍 

Regarding the conversion. You can perhaps save a chunk of runtime by unrolling the loops / folds and just ignoring the rest of the buffer. Try benchmarking with this:

```clojure
(define-private (get-next-random (last-random uint) (ticket-position uint))
	(let ((bytes (sha256 (+ ticket-position last-random)))))
	(+
		(match (element-at bytes u0) byte (buff-to-u8 byte) u0)
		(match (element-at bytes u1) byte (* (buff-to-u8 byte) u256) u0)
		(match (element-at bytes u2) byte (* (buff-to-u8 byte) u65536) u0)
		(match (element-at bytes u3) byte (* (buff-to-u8 byte) u16777216) u0)
		(match (element-at bytes u4) byte (* (buff-to-u8 byte) u4294967296) u0)
		(match (element-at bytes u5) byte (* (buff-to-u8 byte) u1099511627776) u0)
		(match (element-at bytes u6) byte (* (buff-to-u8 byte) u281474976710656) u0)
		(match (element-at bytes u7) byte (* (buff-to-u8 byte) u72057594037927936) u0)
		(match (element-at bytes u8) byte (* (buff-to-u8 byte) u18446744073709551616) u0)
		(match (element-at bytes u9) byte (* (buff-to-u8 byte) u4722366482869645213696) u0)
		(match (element-at bytes u10) byte (* (buff-to-u8 byte) u1208925819614629174706176) u0)
		(match (element-at bytes u11) byte (* (buff-to-u8 byte) u309485009821345068724781056) u0)
		(match (element-at bytes u12) byte (* (buff-to-u8 byte) u79228162514264337593543950336) u0)
		(match (element-at bytes u13) byte (* (buff-to-u8 byte) u20282409603651670423947251286016) u0)
		(match (element-at bytes u14) byte (* (buff-to-u8 byte) u5192296858534827628530496329220096) u0)
		(match (element-at bytes u15) byte (* (buff-to-u8 byte) u1329227995784915872903807060280344576) u0)
	)
)
```

My example takes bytes from the start though, can obviously change that easily.